### PR TITLE
Utilize DML constant input graph node

### PR DESCRIPTION
**Pull request by raoanag**
### Description
This PR also includes,
8b0a55e7cc DML constant pow operator
7520974970 Enable custom heaps based on query-

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 1]
```diff
@@ -293,8 +311,22 @@ namespace DmlGraphFusionHelper
...
+            
+            if (std::holds_alternative<Microsoft::WRL::ComPtr<IDMLOperator>>(nodeInfo.nodeDef))
+            {
+                dmlOperatorGraphNodes[i] = DML_OPERATOR_GRAPH_NODE_DESC{std::get<Microsoft::WRL::ComPtr<IDMLOperator>>(nodeInfo.nodeDef).Get(), nodeInfo.name.data()};
+                dmlGraphNodes[i] = DML_GRAPH_NODE_DESC{DML_GRAPH_NODE_TYPE_OPERATOR, &dmlOperatorGraphNodes[i]};
+            }
+            else
+            {
+                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC_PREVIEW{
+                std::get<std::vector<uint8_t>>(nodeInfo.nodeDef).data(),
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:323]: Params not indented.

Assigning to temporary would help readability.

```
                auto nodeDefinitionData& std::get<std::vector<uint8_t>>(nodeInfo.nodeDef);
                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC_PREVIEW{
                    nodeDefinitionData.data(),
                    nodeDefinitionData.size(),
                    nodeInfo.name.data()
                };
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 2]
```diff
@@ -293,8 +311,22 @@ namespace DmlGraphFusionHelper
...
+                dmlOperatorGraphNodes[i] = DML_OPERATOR_GRAPH_NODE_DESC{std::get<Microsoft::WRL::ComPtr<IDMLOperator>>(nodeInfo.nodeDef).Get(), nodeInfo.name.data()};
+                dmlGraphNodes[i] = DML_GRAPH_NODE_DESC{DML_GRAPH_NODE_TYPE_OPERATOR, &dmlOperatorGraphNodes[i]};
+            }
+            else
+            {
+                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC_PREVIEW{
+                std::get<std::vector<uint8_t>>(nodeInfo.nodeDef).data(), 
+                std::get<std::vector<uint8_t>>(nodeInfo.nodeDef).size(), 
+                nodeInfo.name.data()
+                };
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:326]: Params not indented.

Assigning to temporary would help readability.

```suggestion
                auto nodeDefinitionData& std::get<std::vector<uint8_t>>(nodeInfo.nodeDef);
                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC_PREVIEW{
                    nodeDefinitionData.data(),
                    nodeDefinitionData.size(),
                    nodeInfo.name.data()
                };
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp [hunk 3]
```diff
@@ -289,6 +289,7 @@ namespace Dml::GraphDescBuilder
             std::unordered_map<uint32_t, uint32_t> operatorGraphNodeIndexToMainGraphNodeIndexMap;
             uint32_t graphNodeCount = gsl::narrow_cast<uint32_t>(graphNodes.size());
             const bool isNodeAsOpDesc = graphNodeCreateInfo.nodesAsOperatorDesc.size() > 0;
+            size_t firstOpDescGraphNode = graphNodes.size();
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp:292]: (naming) `firstOpDescGraphNode`? If you were to just look at this variable name and not the type, you'd expect it is a `NodeInfo&` or `DML_OPERATOR_GRAPH_NODE_DESC&`, not an integer.

graphNodeCount?
graphNodeSize?
firstOpDescGraphNodeIndex?

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp [hunk 4]
```diff
@@ -328,21 +331,58 @@ namespace Dml::GraphDescBuilder
...
+
+                            ComPtr<OnnxTensorWrapper> constantInput = constantCpuGraphInputGetter(arg->Name());
+
+                            if (constantInput && constantInput->GetTensorByteSize() < c_maxConstNodeDataSize)
+                            {
+                                std::vector<uint8_t> tensorData;
+                                tensorData.insert(
+                                    tensorData.begin(), 
+                                    static_cast<const uint8_t*>(constantInput->GetData()), 
+                                    static_cast<const uint8_t*>(constantInput->GetData()) + constantInput->GetTensorByteSize());
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp:352]: The constructor has an overload that accepts a range.

```suggestion
                                    auto data = static_cast<const uint8_t*>(constantInput->GetData());
                                    std::vector<uint8_t> tensorData(data, data->constantInput->GetTensorByteSize());
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp (hunk 2 above)
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:326]: The suggested code will also take care of the annoying "Line ends in whitespace" warning :b.

**Review by fdwr — COMMENTED** [review]
Some thoughts.

**Comment by sumitsays** [comment]
Just to be extra cautious, you might want to update this PR by pulling the latest from DmlPrototype branch, because few files have been updated in latest DmlPrototype branch after it got updated with latest main branch.

**Comment by raoanag** [comment]
Need https://github.com/microsoft/onnxruntime/pull/18325 to go in first

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp (hunk 2 above)
Comment (raoanag) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp]: I will also use lint runner on WindowsAI branch, which should handle these warnings for multiple commits at once.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp (hunk 3 above)
Comment (raoanag) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp]: Updated to `firstOpDescGraphNodeIndex`

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 5]
```diff
@@ -288,12 +287,25 @@ namespace DmlGraphFusionHelper
         return initializerPartitionMap;
     }
 
+    enum DML_PREVIEW_OPERATOR_TYPE
+    {
+        DML_PREVIEW_OPERATOR_FIRST = 0xC0000000,
+    };
+
+    struct DML_CONSTANT_DATA_GRAPH_NODE_DESC_PREVIEW
```
Comment (jtseng2) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:295]: What's our strategy to reflect this with new DML1.13 header for the DML_CONSTANT_DATA_GRAPH_NODE_DESC ?

Comment (sumitsays) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:295]: Yes, we should update this as per the new DirectML.h header.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 6]
```diff
@@ -302,8 +302,24 @@
...
+                dmlGraphNodes[i] = DML_GRAPH_NODE_DESC{DML_GRAPH_NODE_TYPE_OPERATOR, &dmlOperatorGraphNodes[i]};
+            }
+            else
+            {
+                auto& nodeDefinitionData = std::get<std::vector<uint8_t>>(nodeInfo.nodeDef);
+	                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC{
+	                    nodeDefinitionData.data(),
+	                    nodeDefinitionData.size(),
+	                    nodeInfo.name.data()
+	                };
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:318]: [nit] indentation

**Review by sumitsays — APPROVED** [review]
:shipit:

**Review by jtseng2 — APPROVED** [review]

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 7]
```diff
@@ -302,8 +302,24 @@
...
+            
+            if (std::holds_alternative<Microsoft::WRL::ComPtr<IDMLOperator>>(nodeInfo.nodeDef))
+            {
+                dmlOperatorGraphNodes[i] = DML_OPERATOR_GRAPH_NODE_DESC{std::get<Microsoft::WRL::ComPtr<IDMLOperator>>(nodeInfo.nodeDef).Get(), nodeInfo.name.data()};
+                dmlGraphNodes[i] = DML_GRAPH_NODE_DESC{DML_GRAPH_NODE_TYPE_OPERATOR, &dmlOperatorGraphNodes[i]};
+            }
+            else
+            {
+                auto& nodeDefinitionData = std::get<std::vector<uint8_t>>(nodeInfo.nodeDef);
+	                dmlConstantGraphNodes[i] = DML_CONSTANT_DATA_GRAPH_NODE_DESC{
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:314]: >	 [](http://example.com/codeflow?start=0&length=1)

Tabs

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp [hunk 8]
```diff
@@ -302,8 +302,24 @@
...
+	                    nodeDefinitionData.data(),
+	                    nodeDefinitionData.size(),
+	                    nodeInfo.name.data()
+	                };
+
+                // TODO: Change as new header is ingested
+                dmlGraphNodes[i] = DML_GRAPH_NODE_DESC{(DML_GRAPH_NODE_TYPE) 2, &dmlConstantGraphNodes[i]};
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionHelper.cpp:321]: >(DML_GRAPH_NODE_TYPE) 2 [](http://example.com/codeflow?start=55&length=23)

Avoid C-style reinterpret casts.

`static_cast<DML_GRAPH_NODE_TYPE>(2)`

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 9]
```diff
@@ -182,6 +182,31 @@
         }
 
         m_isMcdmDevice = (featureLevels.MaxSupportedFeatureLevel == D3D_FEATURE_LEVEL_1_0_CORE_PRIVATE);
+        m_areCustomHeapsSupported = !m_isMcdmDevice;
+
+        if (m_isMcdmDevice) {
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:187]: > { [](http://example.com/codeflow?start=27&length=2)

<-

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 10]
```diff
@@ -182,6 +182,31 @@
...
+                UINT MaxSamplerDescriptorHeapSize;
+                UINT MaxSamplerDescriptorHeapSizeWithStaticSamplers;
+                UINT MaxViewDescriptorHeapSize;
+                _Out_  BOOL ComputeOnlyCustomHeapSupported;
+            } 	D3D12_FEATURE_DATA_D3D12_OPTIONS19;
+
+            D3D12_FEATURE_DATA_D3D12_OPTIONS19 options19 = {};
+
+            // The call may fail in which case the default value is false
+            d3d12Device->CheckFeatureSupport((D3D12_FEATURE) 48 /*D3D12_FEATURE_D3D12_OPTIONS19*/, &options19, sizeof(options19));
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:207]: >(D3D12_FEATURE) 48 [](http://example.com/codeflow?start=45&length=18)

`static_cast<D3D12_FEATURE>(48)`

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp [hunk 11]
```diff
@@ -328,21 +339,59 @@
...
+                            // nodes.  This would allow this size to be reduced while handling the case that 1D scale and zero point
+                            // values that have been de-duplicated with conversion to scalars in kernels.
+                            uint32_t c_maxConstNodeDataSize = 1024 * 1024;
+
+                            ComPtr<OnnxTensorWrapper> constantInput = constantCpuGraphInputGetter(arg->Name());
+
+                            if (constantInput && constantInput->GetTensorByteSize() < c_maxConstNodeDataSize)
+                            {
+                                auto data = static_cast<const uint8_t*>(constantInput->GetData());
+	                            std::vector<uint8_t> tensorData(data, data + constantInput->GetTensorByteSize());
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp:361]: >	 [](http://example.com/codeflow?start=0&length=1)

tab

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp [hunk 12]
```diff
@@ -387,17 +436,28 @@
 
             if (isNodeAsOpDesc)
             {
-                for (auto& opDesc : graphNodeCreateInfo.nodesAsOperatorDesc)
+                for (uint32_t i = 0; i < graphNodeCreateInfo.nodesAsOperatorDesc.size(); ++i)
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp:439]: >uint32_t [](http://example.com/codeflow?start=21&length=8)

Loop counter types should match the loop limit type.

File: onnxruntime/core/providers/dml/OperatorAuthorHelper/MLOperatorAuthorHelper.h [hunk 13]
```diff
@@ -13,6 +13,8 @@
 #define ML_CHECK_BOOL(x) THROW_HR_IF(E_INVALIDARG, !(x))
 #endif
 
+#include <optional>
+
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/OperatorAuthorHelper/MLOperatorAuthorHelper.h:17]: Move up to by top. General headers like anything in `std` should be included before app-specific definitions like `ML_CHECK_BOOL`.

**Review by fdwr — APPROVED** [review]

**Pull request merged**